.travis.yml

```diff
@@ -29,62 +28,53 @@ jobs:
       os: linux
       node_js: 18
       script:
-        - npm run test-unit-cov
+        - npm run coverage
     # test-node14
     - stage: test
       os: linux
       node_js: 14
       script:
-        - npm run test-unit
-        - npm run test-system
+        - npm run test
```

Comment:
I'm pretty sure that's the default script in travis for node, so you should be able to remove the whole `script:`. Or if you want to keep it explicit, you can set it at the top level.

It might also be a good idea to use matrix build and autogenerated test jobs by removing those `stage: test` entries and doing `node_js: [ 18, 16, 14 ], os: [ linux, windows ]`.

Reply:
> I'm pretty sure that's the default script in travis for node, so you should be able to remove the whole script:. Or if you want to keep it explicit, you can set it at the top level.

I would like to keep the script explicit. Puting `script: npm run test` at the top level, defines the default script for the jobs? And it is overwritten if the job provide a `script` field?

> It might also be a good idea to use matrix build and autogenerated test jobs by removing those stage: test entries and doing node_js: [ 18, 16, 14 ], os: [ linux, windows ].

I investigate that at some point but I did not manage to combine stages and job properly. Can you create a job matrix inside the `jobs` top-level field? Something like this?

```yml
 jobs:
  include:
    # lint
    - stage: lint
      script:
        - npm run lint
    # coverage
    - stage: coverage
      os: linux
      node_js: 18
      script:
        - npm run coverage
    # test
    - stage: test
      os: [linux, windows]
      node_js: [14, 16, 18]
      script:
        - npm run test
    # deploy
    - stage: deploy
      if: branch = main
      script:
        - npm run deploy
```

Reply:
> > I'm pretty sure that's the default script in travis for node, so you should be able to remove the whole script:. Or if you want to keep it explicit, you can set it at the top level.
> 
> I would like to keep the script explicit. Puting `script: npm run test` at the top level, defines the default script for the jobs? And it is overwritten if the job provide a `script` field?

Yes, it's the default script for the `test` stage.

> > It might also be a good idea to use matrix build and autogenerated test jobs by removing those stage: test entries and doing node_js: [ 18, 16, 14 ], os: [ linux, windows ].
> 
> I investigate that at some point but I did not manage to combine stages and job properly. Can you create a job matrix inside the `jobs` top-level field? Something like this?

I'm not sure; AFAIR top-level matrix only applies to `test` stage, all the other stages use the first specified environment. So if you do `node_js: [ 18, 16, 14 ], os: [ linux, windows ]` at the top level the tests will be run on all six combinations, while the other stages will use linux-18.

Reply:
I fought for 30 min and could not make this work. I will leave the config file as it is. I agree that creating a matrix is better than enumerating the jobs and I'm interested to make the switch but I don't want to delay this PR any longer.

```yml
arch: amd64
os: linux
dist: focal
language: node_js
cache: npm
node_js: 18

services:
  - mysql

install:
  - npm ci

stages:
  - lint
  - coverage
  - test
  - deploy

jobs:
  include:
    - stage: lint
      script:
        - npm run lint
    - stage: coverage
      os: linux
      node_js: 18
      script:
        - npm run coverage
    - stage: test
      os:
        - linux
        - windows
      node_js:
        - 14
        - 16
        - 18
        - 19
      script:
        - npm run test
    - stage: deploy
      if: branch = main
      script:
        - npm run deploy
```

```
[warn] on jobs.include.os: unexpected sequence, using the first value (linux)
[warn] on jobs.include.node_js: unexpected sequence, using the first value (14)
```

Reply:
That's fine. For what it's worth, what you should have done is do `os: [linux, windows]; node_js: [18, 19, 14, 16]` at the top level — the matrix expansion only expands the test stage, others pick the first (ie. linux-18) by default. I agree that it's a bit confusing :)